### DISCUSSION TITLE
Application-wide services / reactive properties

### DISCUSSION THREAD

**Discussion by drk-mtr** [lit/lit#3126] [Q&A] [answered]
Interested to understand suggested approaches for managing application wide property changes that avoid "prop drilling".

Let's say for example:

- I have a context (configured using @lit-labs/context) that allows me to share an awareness of the current user throughout the app.
- Several nested components that will display elements conditionally, based on whether the user is an admin or not.
- I have a top-nav component that lets me change the current user.

If I change the current user and then load a new page, this works fine. Nested components load elements conditionally based on the current user, because they can get the current user from this context.

However, if I change the user while on a page, nothing updates. This is expected, since child components have no way of knowing the current user has changed.

I know of a few ways to solve this - I could for example bring in an `rxjs` dependency and use `BehaviorSubject`s and subscriptions. Alternatively, I could hold the property on a parent component and pass the updated property all the way down the tree. I'm trying to understand whether there's a best practice "lit way" that avoids either of these.

**Comment by justinfagnani** [comment, marked as answer]
The context protocol allows you to subscribe to updates to the context-provided value.

See the examples from [Consuming a Context](https://github.com/lit/lit/tree/main/packages/labs/context#consuming-a-context) in the context readme.

For `contextProvided` there's a `subscribe` option:

```ts
export class MyElement extends LitElement {
  @contextProvided({context: loggerContext, subscribe: true})
  @property({attribute: false})
  public logger?: Logger;
}
```

And the 4th argument to `ContextConsumer`:
```ts
export class MyElement extends LitElement {
  public logger = new ContextConsumer(
    this,
    loggerContext,
    undefined, // don't need to pass a callback
    true // pass true to get updates if the logger changes
  );
}
```

**Reply by drk-mtr** [reply]
This is really useful, thanks!

I was being a bit silly. I had assumed that this might not be functionality offered by the `context` library, because I wasn't seeing updates despite the `subscribe` option being `true`. However, the mistake I was making was that I was providing a class which was having a property updated, whereas I believe I instead need to be thinking in terms of immutable data structures and actually replacing the object that is being provided.

**Comment by ynxtynxt** [comment]
I am not answering since I don't really have any clue, just that I have been wondering why lit-labs/context wasn't listed on [https://github.com/lit/lit](https://github.com/lit/lit)

And I think from the example,
```js
import {contextRequest} from '@lit-labs/context';
```
Should this be `contextProvided`?

Maybe I should try to answer a bit (but please pardon me for any mistake since I don't really have any clue)
Since I think it's mainly made for [prop drilling](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md) it would only be updating one-way-down, woudn't it? And would the "lit way"
be like firing custom requesting event up to the provider so it can decide to update? (Does that make
any sense? Sorry if it didn't.) Anyway, I am only using Promise for providing anything across the components.

**Reply by drk-mtr** [reply]
Yes I think you're right re. the incorrect import statement.

When you refer to passing the event up, I guess that'd depend on the scenario. You may want to do it so that the component firing the event consumes the context, in which case you could call a method on the context. Alternatively you could fire custom events as you say. You can probably tell I'm not experienced enough to which would be best in what scenario, but nice to have the options :)

**Comment by dbwodlf3** [comment]
You can use capturing event, or just can use dispatch event.
commnly, i just use dispatch update event to components.

and i avoid such cases as much as possible.

root component have just data, and childs components get from root components data.(commonly a array of objects)
so, if root components data is changed then auto changed child components data.

**Reply by drk-mtr** [reply]
Thanks :) In my scenario, my structure would make prop drilling ugly so the `context` library is really useful. e.g. if I have:

```
- Root component (has an awareness of the current logged in user)
  - Nested component layer 1 (doesn't need an awareness of the current user)
    - Nested component layer 2 (doesn't need an awareness of the current user)
      - Nested component layer 3 (needs an awareness of the current user)
```

Here it'd seem wrong to pass the current user into layer 1 and layer 2.
